### PR TITLE
Locking

### PR DESCRIPTION
**Pull request by mthenw**
This can be only merged if https://github.com/serverless/event-gateway-connector/pull/10 is merged.

### CODE REVIEW

File: cmd/connector/main.go [hunk 1]
```diff
@@ -41,41 +46,37 @@ func main() {
...
 		logger.Fatal(wp.Start())
 	}()
 	defer wp.Stop()
 
 	// Server
-	srv := httpapi.ConfigAPI(store)
+	srv := httpapi.ConfigAPI(store, *port)
 	go func() {
-		logger.Debugf("Starting Config API on port: 4002")
+		logger.Debugf("Starting Config API on port: " + strconv.Itoa(*port))
```
Comment (sebito91) [inline cmd/connector/main.go]: `Debugf` is like `Sprintf`, you don't need `strconv` here...

`logger.debugf("Starting Config API on port: %d", *port)`

File: httpapi/server.go [hunk 2]
```diff
@@ -2,21 +2,22 @@ package httpapi
...
-func ConfigAPI(store connection.Service) *http.Server {
+func ConfigAPI(store connection.Service, port int) *http.Server {
 	router := httprouter.New()
 
 	api := &HTTPAPI{Connections: store}
 	api.RegisterRoutes(router)
 
 	handler := &http.Server{
-		Addr:         ":4002",
+		Addr:         ":" + strconv.Itoa(port),
```
Comment (sebito91) [inline httpapi/server.go]: Replace with `fmt.Sprintf(":%d", port)` for clarity.

File: workerpool/pool.go [hunk 3]
```diff
@@ -58,20 +63,41 @@ func (pool *WorkerPool) Start() error {
...
-				pool.jobs[event.Connection.ID] = &job{
-					connection: event.Connection,
-					workers:    make(map[uint]*worker),
+			if event.Type == watcher.Created {
+				// acquire lock
+				mutex := concurrency.NewMutex(pool.session, lockPrefix+string(event.ID))
+				pool.log.Debugw("locking...", "connectionID", event.ID)
+				ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
+				err := mutex.Lock(ctx)
+				cancel()
```
Comment (sebito91) [inline workerpool/pool.go:80]: This should be a defer, no?

File: workerpool/pool.go [hunk 4]
```diff
@@ -58,20 +63,41 @@ func (pool *WorkerPool) Start() error {
...
+				if err != nil {
+					pool.log.Debugw("unable to lock", "error", err, "connectionID", event.ID)
+					continue
+				}
+				pool.log.Debugw("lock acquired", "ID", event.ID)
+
+				// stark job
```
Comment (sebito91) [inline workerpool/pool.go]: `// start job`

File: workerpool/pool.go [hunk 5]
```diff
@@ -58,20 +63,41 @@ func (pool *WorkerPool) Start() error {
 				for _, worker := range job.workers {
 					worker.done <- true
 				}
+
+				if err := job.mutex.Unlock(context.TODO()); err != nil {
+					pool.log.Errorf("unable to unlock", "error", err, "connectionID", job.connection.ID)
```
Comment (sebito91) [inline workerpool/pool.go]: Change to `Errorw` for the key/value context

File: workerpool/pool.go [hunk 6]
```diff
@@ -58,20 +63,41 @@ func (pool *WorkerPool) Start() error {
...
+				}
+				pool.log.Debugf("unlocked", "connectionID", job.connection.ID)
```
Comment (sebito91) [inline workerpool/pool.go]: Same here, change to `Debugw`

File: workerpool/pool.go [hunk 7]
```diff
@@ -58,20 +63,41 @@ func (pool *WorkerPool) Start() error {
...
 			}
 			return nil
 		case event := <-pool.events:
-			count := event.Connection.Source.NumberOfWorkers()
-			if _, ok := pool.jobs[event.Connection.ID]; !ok {
-				pool.jobs[event.Connection.ID] = &job{
-					connection: event.Connection,
-					workers:    make(map[uint]*worker),
+			if event.Type == watcher.Created {
```
Comment (sebito91) [inline workerpool/pool.go:74]: Would we add a case here that would be `watcher.Done` or `watcher.Closed`? That should for a job to unsubscribe altogether.

File: workerpool/pool.go [hunk 8]
```diff
@@ -58,20 +63,41 @@ func (pool *WorkerPool) Start() error {
...
+				// acquire lock
+				mutex := concurrency.NewMutex(pool.session, lockPrefix+string(event.ID))
+				pool.log.Debugw("locking...", "connectionID", event.ID)
+				ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
+				err := mutex.Lock(ctx)
```
Comment (sebito91) [inline workerpool/pool.go:79]: I would swap these two lines... and make it 

```
ctx, cancel := context.WithTimeout(context.Background(), time.Second * 2)
defer cancel()
if err := mutex.Lock(ctx); err != nil {
    	pool.log.Debugw("unable to lock", "error", err, "connectionID", event.ID)
	    continue
}
```

File: workerpool/pool.go [hunk 9]
```diff
@@ -58,20 +63,41 @@ func (pool *WorkerPool) Start() error {
...
+				cancel()
+				if err != nil {
+					pool.log.Debugw("unable to lock", "error", err, "connectionID", event.ID)
+					continue
+				}
+				pool.log.Debugw("lock acquired", "ID", event.ID)
```
Comment (sebito91) [inline workerpool/pool.go]: Make it `connectionID` to keep consistent

File: watcher/watcher.go [hunk 10]
```diff
@@ -61,7 +60,7 @@ func (w *Watcher) Watch() (<-chan *Event, error) {
 			}
 
 			for _, watchEvent := range resp.Events {
-				event := &Event{ID: extractIDFromKey(watchEvent.Kv.Key)}
+				event := &Event{ID: connection.ID(string(watchEvent.Kv.Key))}
```
Comment (sebito91) [inline watcher/watcher.go:63]: You shouldn't need the double-cast here, try converting straight to `connection.ID`.

File: workerpool/pool.go [hunk 11]
```diff
@@ -114,53 +149,39 @@ func newWorker(id uint, log *zap.SugaredLogger, errors chan<- workerError, close
...
 			return
-		case c := <-w.connection:
-			w.log.Debugf("worker %d started job:  %s", w.id, c.ID)
-			if err := w.handleConnection(c); err != nil {
-				w.errors <- workerError{id: w.id, connectionID: c.ID, err: err}
-				return
+
+		default:
+			// perform the actual connection here
+			for i := 0; i < 3; i++ {
```
Comment (sebito91) [inline workerpool/pool.go:161]: Why remove the function call? I'm not precisely sure how this is going to work until we implement the connection part but if this is indeed part of the source interface then I'd suspect we'd end up calling that function.

Let's leave this for now and can revisit when the connector part is ready.

**Review by sebito91 — CHANGES_REQUESTED** [review]
Good job on integrating the lock parts. A few requests for changes, will need to revisit once the connector part is ready.

File: workerpool/pool.go (hunk 3 above)
Comment (mthenw) [reply workerpool/pool.go:80]: Docs:

```
// Canceling this context releases resources associated with it, so code should
// call cancel as soon as the operations running in this Context complete
```

File: workerpool/pool.go (hunk 8 above)
Comment (mthenw) [reply workerpool/pool.go:79]: It cannot be like that because `cancel` has to be called after `Lock`.

File: watcher/watcher.go (hunk 10 above)
Comment (mthenw) [reply watcher/watcher.go:63]: That would be too much magic.

File: workerpool/pool.go (hunk 3 above)
Comment (sebito91) [reply workerpool/pool.go:80]: OK, I guess I'm confused about the `Lock` portion but if you think it fits then let's just leave it.

**Review by sebito91 — APPROVED** [review]

**Pull request merged**